backend/src/main/java/de/grimsi/gameyfin/service/DownloadService.java

```diff
@@ -34,6 +35,26 @@ public String getDownloadFileName(DetectedGame g) {
         return getFilenameWithExtension(path) + ".zip";
     }
 
+    public long getDownloadFileSize(DetectedGame game) {
```

Comment:
I can't see a reason why this method is needed.
There is already a field called `DetectedGame.diskSize` which is calculated when the game is first found in a library scan.
Maybe I'm missing something?

Reply:
There are two reasons why I didn't use diskSize from the DetectedGame.

First is that we cannot use this value for cases where the path of the game is a directory, because the long value doesn't match the actual content length of the .zip file that is created through the ZipOutputStream. That causes the browser to download a corrupted archive. Therefore I needed to make a check if the game is a file or a directory, which is logic that shouldn't be in a Controller since logic should be contained in Services if possible.

The downside is, that this feature doesn't work for .zips that are created through GameyFin since the ZipOutputStream is streamed directly to the ResponseBody and there is no way to get the content length without creating the archive first and then send it to the client, which would cost time.

Second reason is that diskSize of a file **can** change so it wouldn't represent the actual length anymore. It's a rare edge case but if someone for example used TrueNAS and changed the compression level on the array where they put their games, the file size would change but that change is not represented in the database. So I opted for an on demand check when the file is actually downloaded. Since java.nio.file.Files uses the attributes of the filesystem it doesn't cost any performance anyway.

I just changed the code to use the FilesystemService.getSizeOnDisk() method though to avoid duplicated code. :smile:

Reply:
I think the only way to get the exact file size for archived games would be to pre-calculate it by zipping the game files and get the size of the archive.

To prevent edge cases like the one I described for when the filesystem changes, it would also need a watcher service that recognized file system changes and update the diskSize of a game whenever it is notified that its files changed. And to get the proper content length for ZipOutputStreamed games there would have to be a service that runs on a schedule and zip the game to calculate the final size of the .zip file.

But the problem with that would be that the disk will always get activated whenever the schedule runs. On systems like UnRAID that put the hard drives to sleep it would cause the array to wake every time GameyFin runs a schedule and consume more power and cause disk cycles that will degrade the disks.

Therefore I don't think it's a feasible solution either. I personally can live with not having a time estimate and the size of a downloaded archive, since most of my games are compressed to a single file anyway.

Reply:
> And to get the proper content length for ZipOutputStreamed games there would have to be a service that runs on a schedule and zip the game to calculate the final size of the .zip file

This would use a lot of disk resources for the read/write OPs, wouldn't it? Or would you keep it RAM only?
Also, I haven't tried it or checked this part of the code, but what compression does Gameyfin use to zip files? Shouldn't the complete size be able to be calculated when using no compression at all?

Reply:
Gameyfin doesn't use zip compression at all for performance reasons (it would need less bandwidth but would have a much higher CPU load). But zip files do have some overhead, so the zip archive is actually a few bytes bigger than the sum of its contents.
Regarding the calculation: There is already an open feature request to re-calculate the file size if the content of a game changes (currently only possible by removing the game from the library and adding it again by a scan), so maybe there will be a solution in the near future. Although for single files it shouldn't be a problem to calculate the size ad-hoc (it's only slow for directories with a lot of subfolder and files in it).
The current state looks fine to me.